### ISSUE TITLE
Stop hardcoding '99' as the maximum Firefox version ever possible

### ISSUE THREAD

**Issue by diox** [mozilla/addons-server#6016]
In the [version compare code](https://github.com/mozilla/addons-server/blob/ddc439ca0cdf64b6569f36dbdf7c6cd64629e717/src/olympia/versions/compare.py), the [search code](https://github.com/mozilla/addons-server/blob/f74cf659602a1c4f066a537ba52c8d620111cda1/src/olympia/search/views.py#L580-L582) and the [applications definition](https://github.com/mozilla/addons-server/blob/ddc439ca0cdf64b6569f36dbdf7c6cd64629e717/src/olympia/constants/applications.py#L176-L181) the value `99` is hardcoded. It represents the maximum Firefox version ever possible, that value is supposed to be always higher than Firefox's (and other apps).

This results in things like:
```
ipdb> dict_from_int(version_int('99'))
{'pre': None, 'major': 99, 'minor3': 0, 'minor2': 0, 'minor1': 0, 'alpha': None, 'alpha_ver': 0, 'pre_ver': 0}
ipdb> dict_from_int(version_int('99999'))
{'pre': None, 'major': 99, 'minor3': 0, 'minor2': 0, 'minor1': 0, 'alpha': None, 'alpha_ver': 0, 'pre_ver': 0}
```

Breaking some of the compatibility comparison code once Firefox reaches version 99. Firefox not ever reaching version 99 may have been a valid assumption to make [six years ago when it was implemented](https://github.com/mozilla/addons-server/commit/ca6d1a1) but if the naming scheme does not change, that won't be the case in six more years...

**Comment by diox** [comment]
Note: the `AddonIndexer` uses 9999 as the maximum value instead. Currently `version_sidebar()` used to display the app version facets sidebar converts it to `99` and excludes it because it's in the app's `exclude_versions`. When fixing this issue, be careful about that, we don't want to suddenly show "Firefox 9999' in the sidebar as one of the possible versions to filter on...

**Comment by diox** [comment]
We also need to hunt down usage of `min` and `max` on version strings. It works as long as the versions we're considering don't go over 100...

**Comment by diox** [comment]
Keep this one opened, stale bot.

**Comment by eviljeff** [comment]
For extension version strings, Chrome uses 65535 as it's max so we may as well standardize on that.

**Comment by eviljeff** [comment]
Looking into the version string code is a bit of a rabbit hole.  Documenting what I found here.

For a start our version parsing code can't actually handle full toolkit version strings (despite what we claim on MDN).  AMO expects:
 `<number-major>.<number-minor1>.<number-minor2>.<number-minor3><alpha/beta><pre>` where `<alpha/beta>` is "a" or "b" followed by a number; and `<pre>` is "pre" followed by a number.

Toolkit[1] version strings are any number of version-parts, i.e. `<version-part>.<version-part>....` where _each_  version part is `<number-a><string-b><number-c><string-d>`. `56b1c.4pre.1.1.4foo` is technically a valid toolkit version string that AMO wouldn't know how to understand currently.

We replace `*` in the version string with `99` when we're converting it into a single number, which means that `1.100` is > `1.*`.  We _also_ assume that all numbers only have 2 digits in base-10  when converted to a string, so `99.999` is calculated as `999990000200100` but `100` is 100000000200100 - i.e. less.

Chrome extension version strings[2] are `<number-major>.<number-minor1>.<number-minor2>.<number-minor3>`.  So effectively AMO supports a subset of toolkit version strings, which is (thankfully) a superset of the strings Chrome supports for extensions.

[1] https://developer.mozilla.org/en-US/docs/Mozilla/Toolkit_version_format
[2] https://developer.chrome.com/extensions/manifest/version

**Comment by eviljeff** [comment]
Note, Chrome's choice of `<number-major>.<number-minor1>.<number-minor2>.<number-minor3>`  with a maximum of 65535 for each number conveniently allows the version to be encoded as a single number in a 64bit unsigned int.  
If we continue to support the alpha, beta, pre, suffixes then we can't do the same- we'll have to rely on storing the "number" as a string in the database, or support a lower maximum.

**Comment by ioanarusiczki** [comment]
@eviljeff  Hm,  I guess I should test around  `version` or `strict_max_version` from manifest.json but I'm not sure. 
Please update this issue with STR for qa. 
Thanks!

**Comment by eviljeff** [comment]
It's application compatibility, so `strict_max_version` (`version` in the manifest shouldn't be affected)
STR:
- in django admin add some new app versions for 99, 100, 101.
- submit some addons or versions with `max_version` and `min_version`, plus no `max_version` (which should be `*`.
- check the addon update service returns the correct addons

**Comment by eviljeff** [comment]
Also, test with the `appversion` param in the search api https://addons-server.readthedocs.io/en/latest/topics/api/addons.html#search

**Comment by ioanarusiczki** [comment]
@eviljeff @diox 
- I added 99, 100 and 101 in django admin.
- I could upload add-ons which with strict_min_version 99, strict_max101 or no max version.
- I could install and update versions for add-ons with strict_min_version 70 and strict_max_version 99,100 or 101 (I used a lang pack for testing).
- I could install and upload a new version for  an extension with strict_min_version 99 and no max version (or max version 101) by changing the user agent `Mozilla/5.0 (Windows NT 10.0; rv:99.0) Gecko/20100101 Firefox/99.0` (using the responsive design mode). Problem here is that about:addons doesn't work in responsive design mode (at least so far I could not find a solution) so I could not check the update service in this case.
I'll continue to investigate it tomorrow too (as I have today other testing priorities and this is not impacting anything yet) and the [API](https://github.com/mozilla/addons-server/issues/6016#issuecomment-550293581) too and get back with results.
